runtime/doc/lua.txt

```diff
@@ -4433,4 +4433,53 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
         (`string[]`)
 
 
+==============================================================================
+Lua module: vim.net                                                  *vim.net*
+
+vim.net.download({url}, {opts})                           *vim.net.download()*
+    Asynchronously download a file
```

Comment:
A typical pattern is that passing a callback makes it async, and omitting a callback makes it synchronous.

Something to ask is, why is the `as` (aka `file`) option necessary? Is it awkward for the caller to write the result to a file using Nvim's stdlib? If so, why, and can we fix that? I would expect it to be as simple as:

```lua
-- Asyndc
vim.net.request('https://...', {}, function(r)
  vim.fn.writefile(r.data, '/my/file')
end)

-- Sync
local r = vim.net.request('https://...', {})
```

Also: we are currently figuring out a [long-term story for promises/async](https://github.com/neovim/neovim/issues/19624). But until then, we should try to use common patterns. `vim.system()` returns an object that provides `:wait()`. Should we mimic that pattern here? I think I prefer the "optional callback" pattern of libuv/luv, but @lewis6991  may be able to comment on whether the `:wait()` pattern has advantages.

Reply:
> Something to ask is, why is the as (aka file) option necessary?

I haven't thought this thoroughly. The first reason I can think of is that it may not always be desirable to **need** to pass the downloaded file through Neovim to write it. Downloading a big file may cause memory issues (?) (but then, this may be an edge case(?))

Reply:
Advantages of `wait()` is that it:
- makes it slightly easier to type annotate as the object always returns the same type.
- generally more versatile. I.e. you might want to submit multiple tasks and then wait on all of them. It's possible to do this with the other method of providing a dummy callback, but is ugly.

  ```lua
   local a = foo(function() end)
   local b = foo(function() end)
   local c = foo(function() end)

   a:wait()
   b:wait()
   c:cancel()
  ````

I think there is merit in the other pattern, and I don't think we need to restrict ourselves to either one.

I think for `vim.net`, the `wait()` pattern might pan out better longer term. For low level uv functions, the other pattern works well.